pkg/prometheus/promcfg.go

```diff
@@ -43,17 +43,34 @@ var (
 	invalidLabelCharRE = regexp.MustCompile(`[^a-zA-Z0-9_]`)
 )
 
-// ConfigGenerator is used to create Prometheus configurations from operator resources.
+// ConfigGenerator knows how to generate a Prometheus configuration which is
+// compatible with a given Prometheus version.
 type ConfigGenerator struct {
-	logger log.Logger
+	logger  log.Logger
+	version semver.Version
 }
 
-// NewConfigGenerator creates a ConfigGenerator instance using the provided Logger.
-func NewConfigGenerator(logger log.Logger) *ConfigGenerator {
-	cg := &ConfigGenerator{
-		logger: logger,
+// NewConfigGenerator creates a ConfigGenerator for the provided Prometheus resource.
+func NewConfigGenerator(logger log.Logger, p *v1.Prometheus) (*ConfigGenerator, error) {
+	promVersion := operator.StringValOrDefault(p.Spec.Version, operator.DefaultPrometheusVersion)
+	version, err := semver.ParseTolerant(promVersion)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse prometheus version")
+	}
+
+	logger = log.With(logger, "version", promVersion)
```

Comment:
would the logger benefit from having the name of `p` attached here?

Reply:
The logger already has a `key` field inherited from `sync()` which is formatted as `namespace/name`.

https://github.com/prometheus-operator/prometheus-operator/blob/1c913200060b0dbfb2cdb41bbc4956b1437f5678/pkg/prometheus/operator.go#L1220